# `quotationline` — bidder's priced BOQ line items

Unity Catalog: `ingestion_framework_test.bid_data_exploration.quotationline`

Expected: the real equivalent of the per-item CIF/Erection rows the sample data's Excel/PDF parsers currently extract by hand. Look for: item number, description, unit, qty, rate(s), total(s), whether CIF/Erection are separate columns or need to be derived, and the foreign key back to `rfqvendor`.

## Run 1 — initial exploration ✅ *(run in Databricks — awaiting results to document findings)*

In [0]:
%sql
DESCRIBE TABLE ingestion_framework_test.bid_data_exploration.quotationline

col_name,data_type,comment
RFQNUM,varchar(20),null
RFQLINENUM,"decimal(38,10)",null
VENDOR,varchar(44),null
QUOTATIONLINEID,"decimal(38,10)",null
ITEMNUM,varchar(15),null
MANUFACTURER,varchar(44),null
MODELNUM,varchar(100),null
ORDERQTY,"decimal(15,2)",null
ORDERUNIT,varchar(25),null
UNITCOST,"decimal(18,4)",null


In [0]:
%sql
SELECT COUNT(*) AS row_count FROM ingestion_framework_test.bid_data_exploration.quotationline

row_count
1660753


In [0]:
%sql
SELECT * FROM ingestion_framework_test.bid_data_exploration.quotationline LIMIT 20

RFQNUM,RFQLINENUM,VENDOR,QUOTATIONLINEID,ITEMNUM,MANUFACTURER,MODELNUM,ORDERQTY,ORDERUNIT,UNITCOST,LINECOST,EOQ,DELIVERYTIME,DELIVERYDATE,ENTERDATE,ENTERBY,ISAWARDED,SELECTEDFORDISPLAY,GLCREDITACCT,TAX1CODE,TAX1,TAX2CODE,TAX2,TAX3CODE,TAX3,TAX4CODE,TAX4,TAX5CODE,TAX5,CATALOGCODE,MEMO,DESCRIPTION,QUOTESTARTDATE,QUOTEENDDATE,LINECOST2,VENDORPACKCODE,VENDORPACKQUANTITY,VENDORWAREHOUSE,SITEID,ORGID,LINETYPE,ITEMSETID,CONDITIONCODE,COMMODITYGROUP,COMMODITY,LANGCODE,HASLD,MKTPLCITEM,ROWSTAMP,AWARDCOST,ISALTAWARDED,ALTAWARDEDLINE,QL1,QL5,QL4,QL3,SUGAWARD,QL2,CONVERSION,LOADEDCOST,ISEBID,ISPIAWARDED,PRICEIMPACT,BOQITEMNUM,APPLYDISCOUNTAFTERBIDS,DISCOUNT_PERCENT,DISCOUNT_UNITCOST,DISCOUNT_APPLIED_AFTERBIDS,LINECOSTWDIS,QL_EXTRA1,WARRANTYDURATION,WARRANTYPERITEM,WARRANTYREQ
D19885308,4.0000000000,002537,2148023772.0000000000,null,null,null,50.00,NO,325.0000,16250.0000,0.00,null,null,2015-02-15T09:08:17Z,002537S2W0,1.0000000000,1.0000000000,XD-0000-00000-251040-00-000000000-00-000000000,null,0.0000,null,0.00,null,0.00,null,0.00,null,0.00,null,null,MC 75/75A6 Battery with door 3600 mAH (part no. BTRY-KT-1R5X-MC7XR),null,null,null,null,null,null,ADDC,ADDCORG,MATERIAL,null,null,null,null,EN,0E-10,0E-10,6617626984,null,0E-10,null,null,null,null,null,0E-10,QUOTED,null,16250.0000,0E-10,0E-10,0E-10,null,0E-10,0.00,null,0E-10,16250.0000,null,null,null,null
D19885308,5.0000000000,002537,2148023774.0000000000,null,null,null,3.00,NO,215.0000,645.0000,0.00,null,null,2015-02-15T09:08:17Z,002537S2W0,1.0000000000,1.0000000000,XD-0000-00000-251040-00-000000000-00-000000000,null,0.0000,null,0.00,null,0.00,null,0.00,null,0.00,null,215,"MC 75/75A6 Stylus, Spring-Loaded, 3-Pack, tethered (STYLUS-00002-03R)",null,null,null,null,null,null,ADDC,ADDCORG,MATERIAL,null,null,null,null,EN,0E-10,0E-10,6617626985,null,0E-10,null,null,null,null,null,0E-10,QUOTED,null,645.0000,0E-10,0E-10,0E-10,null,0E-10,0.00,null,0E-10,645.0000,null,null,null,null
D19885308,6.0000000000,002537,2148023776.0000000000,null,null,null,5.00,NO,12800.0000,64000.0000,0.00,null,2015-04-05T00:00:00Z,2015-02-15T09:08:17Z,002537S2W0,0E-10,1.0000000000,XD-0000-00000-251040-00-000000000-00-000000000,null,0.0000,null,0.00,null,0.00,null,0.00,null,0.00,null,null,Rugged Laptops,null,2015-04-30T00:00:00Z,null,null,null,null,ADDC,ADDCORG,MATERIAL,null,null,null,null,EN,0E-10,0E-10,6617626986,null,0E-10,null,null,null,null,null,0E-10,QUOTED,null,64000.0000,0E-10,0E-10,0E-10,null,0E-10,0.00,null,0E-10,64000.0000,null,null,null,null
D19885308,1.0000000000,003318,2148023778.0000000000,null,null,null,35.00,NO,328.0000,11480.0000,0.00,null,null,2015-02-04T12:19:12Z,003318JFIZ,0E-10,1.0000000000,XD-0000-00000-251040-00-000000000-00-000000000,null,0.0000,null,0.00,null,0.00,null,0.00,null,0.00,null,null,MC 75/75A6 Snap on Serial Cable (Part No. 25-102776-02R),null,null,null,null,null,null,ADDC,ADDCORG,MATERIAL,null,null,null,null,EN,0E-10,0E-10,6617463221,null,0E-10,null,null,null,null,null,0E-10,QUOTED,null,11480.0000,0E-10,0E-10,0E-10,null,0E-10,0.00,null,0E-10,11480.0000,null,null,null,null
D19885308,2.0000000000,003318,2148023780.0000000000,null,null,null,5.00,NO,328.0000,1640.0000,0.00,null,null,2015-02-04T12:19:12Z,003318JFIZ,0E-10,1.0000000000,XD-0000-00000-251040-00-000000000-00-000000000,null,0.0000,null,0.00,null,0.00,null,0.00,null,0.00,null,null,MC 75/75A6 Snap on USB Cable (Part No. 25-102775-02R),null,null,null,null,null,null,ADDC,ADDCORG,MATERIAL,null,null,null,null,EN,0E-10,0E-10,6617343406,null,0E-10,null,null,null,null,null,0E-10,QUOTED,null,1640.0000,0E-10,0E-10,0E-10,null,0E-10,0.00,null,0E-10,1640.0000,null,null,null,null
D19885308,3.0000000000,003318,2148023782.0000000000,null,null,null,25.00,NO,595.0000,14875.0000,0.00,null,null,2015-02-04T12:19:12Z,003318JFIZ,0E-10,1.0000000000,XD-0000-00000-251040-00-000000000-00-000000000,null,0.0000,null,0.00,null,0.00,null,0.00,null,0.00,null,null,MC 75/75A6 Cigarette Lighter Car Charger (part no. 25-70979-02R),null,null,null,null,null,null,ADDC,ADD

## Run 2 — follow-up queries ⏳ *(pending — not yet run)*

First pass findings (full write-up in `databricks/FINDINGS.md`):
- **`BOQITEMNUM` is a real, dedicated column** — almost certainly the native equivalent of the sample data's hierarchical item numbers (`1.2.3`). Null in this sample (these are catalog/material RFQs, not detailed-BOQ ones) — need to find a populated example.
- **No separate CIF/Erection columns** — just one `UNITCOST`/`LINECOST` pair per line. Leading theory: the `LINETYPE` column (`MATERIAL` vs. `SERVICE` seen in the sample) is how CIF (supply) vs. Erection (install) get split — as two separate line rows sharing the same `BOQITEMNUM`, not two columns on one row. Needs confirming.
- **`ISAWARDED` is tracked per line, per vendor** — confirmed in the sample: for tender `D19885308`, 3 vendors (002537, 003318, 9929192) each quoted the same items, and only vendor 002537's lines 4-5 show `ISAWARDED=1`. This is the real award-decision mechanism, line by line, not just a single winner per RFQ.
- **Correction to my last note in `06_cross_table_relationships`:** there is **no `RFQVENDORID`/`RFQVENDOR_ID` column in `quotationline` at all** — I was wrong to treat the earlier uncommented query as confirming that column exists (outputs were stripped, so it may well have errored rather than succeeded — I shouldn't have inferred success from it being left uncommented). The real join back to `rfqvendor` looks like the **composite key `(RFQNUM, VENDOR)`**, which both tables have.
- Sample shows RFQs with 500+ sequential `RFQLINENUM`s (`D-104880` at line 557-559) — consistent with real large BOQs having ~218 rows per lot like the sample data.
- A third org appears (`TRANS`/`TRANSORG`) beyond `ADWEA`/`ADDC` — reinforces that this dataset spans the whole group, not just ADDC.

### Find a populated `BOQITEMNUM` example (a real detailed-BOQ tender)

In [0]:
%sql
SELECT RFQNUM, VENDOR, BOQITEMNUM, RFQLINENUM, DESCRIPTION, ORDERQTY, UNITCOST, LINECOST, LINETYPE
FROM ingestion_framework_test.bid_data_exploration.quotationline
WHERE BOQITEMNUM IS NOT NULL
ORDER BY ENTERDATE DESC
LIMIT 30

RFQNUM,VENDOR,BOQITEMNUM,RFQLINENUM,DESCRIPTION,ORDERQTY,UNITCOST,LINECOST,LINETYPE
D19892718,001597,3001013.80,2.0000000000,"Fuse HHD 80A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,675.0000,6075.0000,MATERIAL
D19892718,001597,3000213.50,1.0000000000,"Fuse HHD 50A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,625.0000,5625.0000,MATERIAL
D19892718,001597,3001013.125,3.0000000000,"Fuse HHD 125A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,715.0000,6435.0000,MATERIAL
D19892718,001597,3001813.160,4.0000000000,"Fuse HHD 160A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,950.0000,8550.0000,MATERIAL
D19892718,99678575,3001013.125,3.0000000000,"Fuse HHD 125A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,423.5300,3811.7700,MATERIAL
D19892718,99678575,3001813.160,4.0000000000,"Fuse HHD 160A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,541.1800,4870.6200,MATERIAL
D19892718,99678575,3000213.50,1.0000000000,"Fuse HHD 50A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,305.8900,2753.0100,MATERIAL
D19892718,99678575,3001013.80,2.0000000000,"Fuse HHD 80A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,423.5300,3811.7700,MATERIAL
D19892718,99687643,3001013.125,3.0000000000,"Fuse HHD 125A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,469.5000,4225.5000,MATERIAL
D19892718,99687643,3001013.80,2.0000000000,"Fuse HHD 80A BU 3/7,2kV e=192mm Brand : Siba Origin : Germany HS Code : 85351000",9.00,407.5000,3667.5000,MATERIAL


### Is `LINETYPE` the CIF/Erection split? (distribution first)

In [0]:
%sql
SELECT LINETYPE, COUNT(*) AS n FROM ingestion_framework_test.bid_data_exploration.quotationline
GROUP BY LINETYPE ORDER BY n DESC

LINETYPE,n
SERVICE,1032113
ITEM,351248
MATERIAL,277304
STDSERVICE,78
TOOL,10


### Does the same `BOQITEMNUM` appear twice per vendor with different `LINETYPE`s?
If CIF/Erection are split into two rows, expect `COUNT(*) = 2` for most groups below.

In [0]:
%sql
SELECT RFQNUM, VENDOR, BOQITEMNUM, COUNT(*) AS n, COLLECT_SET(LINETYPE) AS linetypes
FROM ingestion_framework_test.bid_data_exploration.quotationline
WHERE BOQITEMNUM IS NOT NULL
GROUP BY RFQNUM, VENDOR, BOQITEMNUM
ORDER BY n DESC
LIMIT 30

RFQNUM,VENDOR,BOQITEMNUM,n,linetypes
N-16578,002321,PART 1,117,List(SERVICE)
N-16578,001456,PART 1,117,List(SERVICE)
N-16578,001938,PART 1,117,List(SERVICE)
N-19339,003155,NUMBER,49,List(MATERIAL)
N-19339,9913162,NUMBER,49,List(MATERIAL)
N-19339,99456349,NUMBER,49,List(MATERIAL)
N-19339,99471391,NUMBER,49,List(MATERIAL)
N-19339,99110942,NUMBER,49,List(MATERIAL)
N-17334,001635,LOT 01,13,"List(SERVICE, MATERIAL)"
N-17334,001635,LOT 03,12,"List(SERVICE, MATERIAL)"


### Which RFQs actually look like real detailed BOQs (100+ line items)?

In [0]:
%sql
SELECT RFQNUM, COUNT(*) AS line_count, COUNT(DISTINCT VENDOR) AS vendor_count
FROM ingestion_framework_test.bid_data_exploration.quotationline
GROUP BY RFQNUM
HAVING COUNT(*) > 100
ORDER BY line_count DESC
LIMIT 30

RFQNUM,line_count,vendor_count
N-20585,52398,9
N-18709,42336,9
N-19899,38740,5
N-19304,33824,7
N-18530,33312,6
D-104888,28544,16
D-104881,28544,16
N-18786,27480,5
D-104880,19624,11
D-104801,17840,10


### Does D-111808 (or similar) exist directly in this table?

In [0]:
%sql
SELECT RFQNUM, VENDOR, BOQITEMNUM, DESCRIPTION, ORDERQTY, UNITCOST, LINECOST, LINETYPE, ISAWARDED
FROM ingestion_framework_test.bid_data_exploration.quotationline
WHERE RFQNUM LIKE 'D-111808%'
ORDER BY VENDOR, RFQLINENUM
LIMIT 50

RFQNUM,VENDOR,BOQITEMNUM,DESCRIPTION,ORDERQTY,UNITCOST,LINECOST,LINETYPE,ISAWARDED
D-111808,001565,null,"Construction contract for the Replacement of Shobaisi, Samha and DRA Primary Substation in Eastern Region",1.00,234651862.0000,234651862.0000,MATERIAL,0E-10
D-111808,001938,null,"Construction contract for the Replacement of Shobaisi, Samha and DRA Primary Substation in Eastern Region",1.00,142459514.1000,142459514.1000,MATERIAL,1.0000000000
D-111808,99102876,null,"Construction contract for the Replacement of Shobaisi, Samha and DRA Primary Substation in Eastern Region",1.00,208545844.6400,208545844.6400,MATERIAL,0E-10
D-111808,99107175,null,"Construction contract for the Replacement of Shobaisi, Samha and DRA Primary Substation in Eastern Region",1.00,195458685.0000,195458685.0000,MATERIAL,0E-10
D-111808,99443592,null,"Construction contract for the Replacement of Shobaisi, Samha and DRA Primary Substation in Eastern Region",1.00,221525111.0000,221525111.0000,MATERIAL,0E-10
D-111808,99456349,null,"Construction contract for the Replacement of Shobaisi, Samha and DRA Primary Substation in Eastern Region",1.00,161562485.0800,161562485.0800,MATERIAL,0E-10
D-111808,99473989,null,"Construction contract for the Replacement of Shobaisi, Samha and DRA Primary Substation in Eastern Region",1.00,179400000.0000,179400000.0000,MATERIAL,0E-10
D-111808,9958404,null,"Construction contract for the Replacement of Shobaisi, Samha and DRA Primary Substation in Eastern Region",1.00,149332885.0000,149332885.0000,MATERIAL,0E-10


### Confirm the (RFQNUM, VENDOR) join back to `rfqvendor` actually lines up

In [0]:
%sql
SELECT ql.RFQNUM, ql.VENDOR, COUNT(*) AS line_count, SUM(ql.LINECOST) AS total_linecost
FROM ingestion_framework_test.bid_data_exploration.quotationline ql
GROUP BY ql.RFQNUM, ql.VENDOR
ORDER BY line_count DESC
LIMIT 20

RFQNUM,VENDOR,line_count,total_linecost
N-18610,9913229,8240,54604215.0000
N-18610,9912598,8240,44042018.0000
N-19899,99473289,7748,147524200.2000
N-19899,001033,7748,119166753.6000
N-19899,9943988,7748,92924091.8200
N-19899,99506197,7748,113063110.0000
N-19899,99515572,7748,90806060.0000
N-20585,99101236,5822,3544740.0000
N-20585,002012,5822,178242560.0000
N-20585,99132201,5822,96557659.4000


**Observations (updated after first real run):**
- `BOQITEMNUM` is the real hierarchical item number column — needs a populated example to confirm format matches `1.2.3`-style numbering.
- No CIF/Erection columns — `LINETYPE` (`MATERIAL`/`SERVICE`) splitting into two rows per `BOQITEMNUM` is the leading theory, unconfirmed.
- `ISAWARDED` genuinely tracks award status per vendor per line — this is real, granular award data, richer than anything in the sample model.
- Join back to `rfqvendor` is the composite key `(RFQNUM, VENDOR)`, **not** a single `RFQVENDORID`/`RFQVENDOR_ID` FK column — that column does not exist in this table.